File: backend/src/services/doctorService.js
import { Op } from 'sequelize';
import Doctor from '../models/Doctor.js';
import User from '../models/User.js';
import redisClient from '../config/redis.js';
import logger from '../config/logger.js';

/**
 * Generate Redis cache key for doctor search
 */
const generateCacheKey = (filters, page, limit, sortBy, sortOrder) => {
  const filterString = JSON.stringify(filters);
  return `doctor_search:${filterString}:${page}:${limit}:${sortBy}:${sortOrder}`;
};

/**
 * Search doctors with filters and pagination
 */
export const searchDoctors = async (filters, page = 1, limit = 20, sortBy = 'created_at', sortOrder = 'DESC') => {
  try {
    const cacheKey = generateCacheKey(filters, page, limit, sortBy, sortOrder);
    
    // Check Redis cache
    const cachedResults = await redisClient.get(cacheKey);
    if (cachedResults) {
      logger.info('Doctor search results retrieved from cache');
      return JSON.parse(cachedResults);
    }

    // Build where clause
    const whereClause = {};

    // Free-text search across multiple fields
    if (filters.searchQuery) {
      whereClause[Op.or] = [
        { specialty: { [Op.iLike]: `%${filters.searchQuery}%` } },
        { sub_specialty: { [Op.iLike]: `%${filters.searchQuery}%` } },
        { office_city: { [Op.iLike]: `%${filters.searchQuery}%` } },
        { office_state: { [Op.iLike]: `%${filters.searchQuery}%` } },
        { bio: { [Op.iLike]: `%${filters.searchQuery}%` } }
      ];
    }

    if (filters.specialty) {
      whereClause.specialty = { [Op.iLike]: `%${filters.specialty}%` };
    }

    if (filters.subSpecialty) {
      whereClause.sub_specialty = { [Op.iLike]: `%${filters.subSpecialty}%` };
    }

    if (filters.city) {
      whereClause.office_city = { [Op.iLike]: `%${filters.city}%` };
    }

    if (filters.state) {
      whereClause.office_state = { [Op.iLike]: `%${filters.state}%` };
    }

    if (filters.zipCode) {
      whereClause.office_zip_code = filters.zipCode;
    }

    if (filters.minFee !== undefined || filters.maxFee !== undefined) {
      whereClause.consultation_fee = {
        [Op.between]: [
          filters.minFee || 0,
          filters.maxFee || 999999
        ]
      };
    }

    if (filters.languages && filters.languages.length > 0) {
      whereClause.languages_spoken = {
        [Op.contains]: filters.languages
      };
    }

    if (filters.isAcceptingPatients !== undefined) {
      whereClause.is_accepting_patients = filters.isAcceptingPatients;
    }

    if (filters.telehealthEnabled !== undefined) {
      whereClause.telehealth_enabled = filters.telehealthEnabled;
    }

    // Calculate offset
    const offset = (page - 1) * limit;

    // Query doctors with pagination
    const { count, rows: doctors } = await Doctor.findAndCountAll({
      where: whereClause,
      include: [
        {
          model: User,
          as: 'user',
          attributes: ['id', 'first_name', 'last_name', 'email', 'phone', 'profile_image']
        }
      ],
      order: [[sortBy, sortOrder]],
      limit: parseInt(limit),
      offset: offset,
      distinct: true
    });

    // Calculate pagination metadata
    const totalPages = Math.ceil(count / limit);
    const result = {
      doctors,
      pagination: {
        total: count,
        page: parseInt(page),
        limit: parseInt(limit),
        totalPages
      }
    };

    // Cache results for 5 minutes
    await redisClient.setEx(cacheKey, 300, JSON.stringify(result));
    logger.info(`Doctor search executed, ${count} results found`);

    return result;
  } catch (error) {
    logger.error('Error searching doctors:', error);
    throw error;
  }
};

/**
 * Get doctor by ID with associations
 */
export const getDoctorById = async (doctorId) => {
  try {
    const cacheKey = `doctor:${doctorId}`;
    
    // Check Redis cache
    const cachedDoctor = await redisClient.get(cacheKey);
    if (cachedDoctor) {
      logger.info(`Doctor ${doctorId} retrieved from cache`);
      return JSON.parse(cachedDoctor);
    }

    const doctor = await Doctor.findByPk(doctorId, {
      include: [
        {
          model: User,
          as: 'user',
          attributes: ['id', 'first_name', 'last_name', 'email', 'phone', 'profile_image', 'email_verified']
        }
      ]
    });

    if (!doctor) {
      return null;
    }

    // Cache for 5 minutes
    await redisClient.setEx(cacheKey, 300, JSON.stringify(doctor));
    logger.info(`Doctor ${doctorId} retrieved from database`);

    return doctor;
  } catch (error) {
    logger.error(`Error getting doctor ${doctorId}:`, error);
    throw error;
  }
};

/**
 * Get doctor availability
 */
export const getDoctorAvailability = async (doctorId) => {
  try {
    const doctor = await Doctor.findByPk(doctorId, {
      attributes: ['id', 'working_hours', 'is_accepting_patients', 'telehealth_enabled']
    });

    if (!doctor) {
      return null;
    }

    return {
      workingHours: doctor.working_hours,
      isAcceptingPatients: doctor.is_accepting_patients,
      telehealthEnabled: doctor.telehealth_enabled
    };
  } catch (error) {
    logger.error(`Error getting doctor availability for ${doctorId}:`, error);
    throw error;
  }
};

/**
 * Invalidate doctor cache
 */
export const invalidateDoctorCache = async (doctorId) => {
  try {
    const cacheKey = `doctor:${doctorId}`;
    await redisClient.del(cacheKey);
    
    // Also invalidate search caches (this is simplified; in production, might use cache tags)
    const searchKeys = await redisClient.keys('doctor_search:*');
    if (searchKeys.length > 0) {
      await redisClient.del(searchKeys);
    }
    
    logger.info(`Cache invalidated for doctor ${doctorId}`);
  } catch (error) {
    logger.error(`Error invalidating cache for doctor ${doctorId}:`, error);
  }
};

File: mobile/lib/services/doctor_service.dart
import '../models/doctor_model.dart';
import '../models/doctor_search_filter.dart';
import 'api_service.dart';

class ApiResponse<T> {

  ApiResponse({
    required this.success,
    this.message,
    this.data,
    this.error,
  });
  final bool success;
  final String? message;
  final T? data;
  final dynamic error;
}

class PaginationMetadata {

  PaginationMetadata({
    required this.total,
    required this.page,
    required this.limit,
    required this.totalPages,
  });

  factory PaginationMetadata.fromJson(Map<String, dynamic> json) => PaginationMetadata(
      total: (json['total'] ?? 0) as int,
      page: (json['page'] ?? 1) as int,
      limit: (json['limit'] ?? 20) as int,
      totalPages: (json['totalPages'] ?? json['total_pages'] ?? 0) as int,
    );
  final int total;
  final int page;
  final int limit;
  final int totalPages;
}

class DoctorSearchResult {

  DoctorSearchResult({
    required this.doctors,
    required this.pagination,
  });
  final List<Doctor> doctors;
  final PaginationMetadata pagination;
}

class DoctorService {

  DoctorService(this._apiService);
  final ApiService _apiService;

  /// Search doctors with filters and pagination
  Future<ApiResponse<DoctorSearchResult>> searchDoctors(
    DoctorSearchFilter filter, {
    int page = 1,
    int limit = 20,
  }) async {
    try {
      // Build query parameters
      final queryParams = filter.toQueryParams();
      queryParams['page'] = page.toString();
      queryParams['limit'] = limit.toString();

      final response = await _apiService.get(
        '/doctors/search',
        queryParams: queryParams,
      );

      if (response.success && response.data != null) {
        final jsonData = response.data as Map<String, dynamic>;
        
        // Parse doctors list
        final doctorsJson = (jsonData['data']['doctors'] ?? []) as List<dynamic>;
        final doctors = doctorsJson
            .map((json) => Doctor.fromJson(json as Map<String, dynamic>))
            .toList();

        // Parse pagination
        final paginationJson = (jsonData['data']['pagination'] ?? {}) as Map<String, dynamic>;
        final pagination = PaginationMetadata.fromJson(paginationJson);

        final result = DoctorSearchResult(
          doctors: doctors,
          pagination: pagination,
        );

        return ApiResponse(
          success: true,
          message: jsonData['message'] as String?,
          data: result,
        );
      } else {
        return ApiResponse(
          success: false,
          message: response.message ?? 'Failed to search doctors',
          error: response.error,
        );
      }
    } catch (e) {
      return ApiResponse(
        success: false,
        message: 'Network error: ${e.toString()}',
        error: e,
      );
    }
  }

  /// Get doctor by ID
  Future<ApiResponse<Doctor>> getDoctorById(String doctorId) async {
    try {
      final response = await _apiService.get('/doctors/$doctorId');

      if (response.success && response.data != null) {
        final jsonData = response.data as Map<String, dynamic>;
        final doctor = Doctor.fromJson(jsonData['data'] as Map<String, dynamic>);

        return ApiResponse(
          success: true,
          message: jsonData['message'] as String?,
          data: doctor,
        );
      } else if (response.statusCode == 404) {
        return ApiResponse(
          success: false,
          message: 'Doctor not found',
          error: 'NOT_FOUND',
        );
      } else {
        return ApiResponse(
          success: false,
          message: response.message ?? 'Failed to fetch doctor',
          error: response.error,
        );
      }
    } catch (e) {
      return ApiResponse(
        success: false,
        message: 'Network error: ${e.toString()}',
        error: e,
      );
    }
  }

  /// Get doctor availability
  Future<ApiResponse<Map<String, dynamic>>> getDoctorAvailability(
      String doctorId) async {
    try {
      final response = await _apiService.get('/doctors/$doctorId/availability');

      if (response.success && response.data != null) {
        final jsonData = response.data as Map<String, dynamic>;
        final availability = jsonData['data'] as Map<String, dynamic>;

        return ApiResponse(
          success: true,
          message: jsonData['message'] as String?,
          data: availability,
        );
      } else if (response.statusCode == 404) {
        return ApiResponse(
          success: false,
          message: 'Doctor not found',
          error: 'NOT_FOUND',
        );
      } else {
        return ApiResponse(
          success: false,
          message: response.message ?? 'Failed to fetch availability',
          error: response.error,
        );
      }
    } catch (e) {
      return ApiResponse(
        success: false,
        message: 'Network error: ${e.toString()}',
        error: e,
      );
    }
  }
}
